protocol/x/clob/keeper/deleveraging.go

```diff
@@ -188,6 +188,50 @@ func (k Keeper) CanDeleverageSubaccount(
 	return false, clobPair.Status == types.ClobPair_STATUS_FINAL_SETTLEMENT, nil
 }
 
+// GateWithdrawalsIfNegativeTncSubaccountSeen gates withdrawals if a negative TNC subaccount exists.
+// It does this by inserting a zero-fill deleveraging operation into the operations queue iff any of
+// the provided liquidatable subaccounts are negative TNC.
+func (k Keeper) GateWithdrawalsIfNegativeTncSubaccountSeen(
+	ctx sdk.Context,
+	liquidatableSubaccountIds []satypes.SubaccountId,
+) (err error) {
+	foundNegativeTncSubaccount := false
+	var negativeTncSubaccountId satypes.SubaccountId
+	for _, subaccountId := range liquidatableSubaccountIds {
+		bigNetCollateral,
+			_,
+			_,
+			err := k.subaccountsKeeper.GetNetCollateralAndMarginRequirements(
+			ctx,
+			satypes.Update{SubaccountId: subaccountId},
+		)
+		if err != nil {
+			return err
+		}
+
+		// If the subaccount has negative TNC, mark that a negative TNC subaccount was found.
+		if bigNetCollateral.Sign() == -1 {
+			foundNegativeTncSubaccount = true
+			negativeTncSubaccountId = subaccountId
+			break
+		}
+	}
+
+	if !foundNegativeTncSubaccount {
+		return nil
```

Comment:
Planning to add more metrics here.
- [ ] Add counters for total calls to this method.
- [ ] Add counters for halted withdrawals (can add alerts on this).
- [ ] Add latency metrics to this method.